packages/cli/src/auth/index.ts

```diff
@@ -1,86 +1,243 @@
+      switch (pathname) {
+        case "/auth/callback": {
+          if (query.token) {
+            const token = Array.isArray(query.token) ? query.token[0] : query.token;
+            const theirState = Array.isArray(query.state)
+              ? query.state[0]
+              : query.state;
+
+            // Check if the payload sent back is the same as the one we sent.
+            if (theirState !== JSON.stringify(loggedIn)) {
+              res.writeHead(400, { "Content-Type": "text/plain" }); // send a 400 response
+              res.end("Unauthorized request, state mismatch", () => {
+                finish(new Error("Unauthorized request, state mismatch"));
+              });
+              reject(
+                new Error(chalk.red("Unauthorized request, state mismatch")),
+              );
+            } else {
+              // Save the token to the config file.
+              const configJson = JSON.parse(fs.readFileSync(credsConfigPath, "utf-8"));
+              fs.writeFileSync(credsConfigPath, JSON.stringify({
+                ...configJson,
+                token
+              }), {
+                encoding: "utf8",
+                mode: 0o600,
+              });
+              res.end(() => {
+                finish();
+              });
+              logger.info(chalk.green(`Successfully logged in.`));
+              resolve(token); // resolve promise with secretKey
+            }
+          } else {
+            res.writeHead(400, { "Content-Type": "text/plain" }); // send a 400 response
+            res.end("No authToken received", () => {
+              finish(new Error("No authToken received"));
+            });
+            reject(new Error(chalk.red("No authToken received")));
+          }
+        }
+      }
     });
 
-    try {
-      await validateKey(response.apiSecretKey);
-    } catch (error) {
-      console.log(error);
-      process.exit(1);
-    }
-    cache.set("api-secret-key", response.apiSecretKey);
-    return response.apiSecretKey;
-  } catch (error) {
-    console.log(error);
+    server.listen(8976);
+  });
+  if (props?.browser) {
+    ora(`Opening a link in your default browser: ${urlToOpen}\n`).start();
+    // Adding this timeout since it feels weird for the browser to open before the spinner.
+    setTimeout(async () => {
+      await open(urlToOpen);
+    }, 2000);
+  } else {
+    logger.info(`Visit this link to authenticate: ${urlToOpen}`);
+  }
+
+  return Promise.race([timerPromise, loginPromise]);
+};
+
+async function getOrCreatePassword(credsConfigPath: string): Promise<string> {
+  const configExists = fs.existsSync(credsConfigPath);
+  const configJson = JSON.parse(fs.readFileSync(credsConfigPath, "utf-8"));
+
+  if (configExists && configJson.password) {
+    return configJson.password;
   }
+
+  const response = await prompts({
+    type: "invisible",
+    name: "password",
+    message: `Please enter a password to authenticate with the CLI`,
```

Comment:
when will this get prompted? when user creates a wallet first time or when unlocking? can we make the message clearer?